### PR TITLE
Python 3.10

### PR DESCRIPTION
**Pull request by esc**
As title

### CODE REVIEW

File: llvmlite/binding/ffi.py [hunk 1]
```diff
@@ -183,7 +183,8 @@ def __call__(self, *args, **kwargs):
 for _lib_path in _lib_paths:
     try:
         lib = ctypes.CDLL(_lib_path)
-    except OSError:
+    except OSError as e:
+        print(e)
```
Comment (sklam) [inline llvmlite/binding/ffi.py:187]: Is this leftover from debugging?

Comment (esc) [reply llvmlite/binding/ffi.py:187]: ok, that makes sense

**Review by sklam — COMMENTED** [review]
I've manually tested the failed to find libllvmlite error and it looks good.

**Comment by mbargull** [comment]
xref from conda-forge where the CPython 3.10 migration has been kicked off yesterday:
https://github.com/conda-forge/llvmlite-feedstock/pull/57
(I.e., if you'd want to test with 3.10 you could (temporarily or additionally) use conda-forge's channel on the CI.)

With https://github.com/regro-cf-autotick-bot/llvmlite-feedstock/blob/846ab90f04df333844eea2d0411c79e4858b59fd/recipe/run_test.py everything apart from macOS CI passed.
The failure observed on macOS is:
```python
===== testing package: llvmlite-0.37.0-py310h003a20c_1 =====
running run_test.py
..s..............Fatal Python error: Segmentation fault

Current thread 0x000000010effddc0 (most recent call first):
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/site-packages/llvmlite/binding/ffi.py", line 151 in __call__
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/site-packages/llvmlite/binding/executionengine.py", line 92 in finalize_object
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/site-packages/llvmlite/tests/test_binding.py", line 622 in get_sum
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/site-packages/llvmlite/tests/test_binding.py", line 773 in test_object_cache_getbuffer
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/unittest/case.py", line 549 in _callTestMethod
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/unittest/case.py", line 591 in run
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/unittest/case.py", line 650 in __call__
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/unittest/suite.py", line 122 in run
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/unittest/suite.py", line 84 in __call__
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/unittest/suite.py", line 122 in run
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/unittest/suite.py", line 84 in __call__
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/unittest/suite.py", line 122 in run
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/unittest/suite.py", line 84 in __call__
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/unittest/runner.py", line 176 in run
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/unittest/main.py", line 271 in runTests
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/site-packages/llvmlite/tests/customize.py", line 97 in run_tests_real
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/site-packages/llvmlite/tests/customize.py", line 112 in runTests
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/unittest/main.py", line 101 in __init__
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/site-packages/llvmlite/tests/customize.py", line 38 in __init__
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/site-packages/llvmlite/tests/__init__.py", line 50 in run_tests
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/site-packages/llvmlite/tests/__init__.py", line 56 in main
  File "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/test_tmp/run_test.py", line 11 in <module>
/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/test_tmp/conda_test_runner.sh: line 3: 27203 Segmentation fault: 11  "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/bin/python" -s "/Users/runner/miniforge3/conda-bld/llvmlite_1635828737204/test_tmp/run_test.py"
Tests failed for llvmlite-0.37.0-py310h003a20c_1.tar.bz2 - moving package to /Users/runner/miniforge3/conda-bld/broken
```
Logs currently available at https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=401139&view=logs&j=9c5ef928-2cd6-52e5-dbe6-9d173a7d951b&t=20c71c51-4b27-578b-485d-06ade2de1d00

**Comment by esc** [comment]
@mbargull thank you for the ping. I have not yet seen such a segfault before during my testing, neither on public CI nor on the Anaconda internal build farm. It will probably make sense to try to set something up to reproduce this locally.

**Comment by esc** [comment]
https://github.com/numba/llvmlite/pull/789 is now provided to circumvent the CI issues present on this PR.

**Comment by esc** [comment]
Oh wow, it went green for the first time, yay!

**Comment by esc** [comment]
Now running on the Anconda internal build farm as: `llvmlite_119`.

**Comment by esc** [comment]
This passed on the anaconda internal build farm:

Screen Shot 2021-11-18 at 22 33 14

**Comment by jakirkham** [comment]
Matti tracked the issue Marcel noted above to the finalization line here (622):

https://github.com/numba/llvmlite/blob/d77dc1bcdb5af040c549f4d1ceeb4db7c8d08718/llvmlite/tests/test_binding.py#L621-L625

More details in comment ( https://github.com/conda-forge/llvmlite-feedstock/pull/58#issuecomment-973405364 )

**Comment by esc** [comment]
> Matti tracked the issue Marcel noted above to the finalization line here (622):
> 
> https://github.com/numba/llvmlite/blob/d77dc1bcdb5af040c549f4d1ceeb4db7c8d08718/llvmlite/tests/test_binding.py#L621-L625
> 
> More details in comment ( [conda-forge/llvmlite-feedstock#58 (comment)](https://github.com/conda-forge/llvmlite-feedstock/pull/58#issuecomment-973405364) )

@jakirkham thank you for posting this. It will probably make sense to debug this during the RC period when we have also built our binary artifacts.

**Comment by mbargull** [comment]
See https://github.com/conda-forge/llvmlite-feedstock/pull/58#issuecomment-973829079 : Building works fine in conda-forge's setup but testing with `conda-forge::python=3.10` fails.

**Comment by mbargull** [comment]
> See [conda-forge/llvmlite-feedstock#58 (comment)](https://github.com/conda-forge/llvmlite-feedstock/pull/58#issuecomment-973829079) : Building works fine in conda-forge's setup but testing with `conda-forge::python=3.10` fails.

@esc , this could/would also mean that you can probably test it in an environment created with your build artifact and `conda-forge::python=3.10`.

**Comment by mbargull** [comment]
@esc, as an update: the segfault on osx-64 for conda-forge stems from the CPython 3.10 plus libffi 3.4 combination.
@jakirkham [found](https://github.com/conda-forge/python-feedstock/issues/522#issuecomment-973896170) and upstream issue for this: https://bugs.python.org/issue44556
I opened https://github.com/conda-forge/python-feedstock/pull/523 for us to handle this.

----
TL;DR looks like `llvmlite=0.37.0` works fine with `python=3.10` overall and the one known encountered issue doesn't stem from llvmlite's code base :).

**Comment by esc** [comment]
> @esc, as an update: the segfault on osx-64 for conda-forge stems from the CPython 3.10 plus libffi 3.4 combination. @jakirkham [found](https://github.com/conda-forge/python-feedstock/issues/522#issuecomment-973896170) and upstream issue for this: https://bugs.python.org/issue44556 I opened [conda-forge/python-feedstock#523](https://github.com/conda-forge/python-feedstock/pull/523) for us to handle this.
> 
> TL;DR looks like `llvmlite=0.37.0` works fine with `python=3.10` overall and the one known encountered issue doesn't stem from llvmlite's code base :).

Great finds! Thank you for keeping us in the loop. It is much appreciated!

**Comment by esc** [comment]
@mbargull would you recommend we pin `libffi` in our recpie?

**Comment by jakirkham** [comment]
No the issue was in how CPython was doing the build itself. We fixed that in conda-forge ( https://github.com/conda-forge/python-feedstock/pull/525 ). Apparently the Python devs also fixed this, but it didn't make it into 3.10.0 ( https://bugs.python.org/issue45350 ).

**Comment by mbargull** [comment]
Yep, what John wrote! All good from our site now. I'll open an issue on Anaconda's recipe repo too so we don't run into the same issue for the `defaults` channel once `defaults::python` gets rebuild with `libffi=3.4`.

**Comment by esc** [comment]
@sklam @stuartarchibald @seibert I think this one is ready, needs a final review, I believe.

**Comment by esc** [comment]
> Yep, what John wrote! All good from our side now. I'll open an issue on Anaconda's recipe repo (\edit: [AnacondaRecipes/python-feedstock#50](https://github.com/AnacondaRecipes/python-feedstock/issues/50) ) too so we don't run into the same issue for the `defaults` channel if `defaults::python=3.10.0` gets rebuild with `libffi=3.4`.

Thank you, much appreciated!! 🙏

**Review by gmarkall — APPROVED** [review]
It looks like all the issues brought up in the discussion are now resolved, so this looks good to me.

**Comment by esc** [comment]
Thank you all for the reviews!
